src/taipy/gui/renderers/_class_api/element_api.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 Avaiga Private Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+from __future__ import annotations
+
+import typing as t
+from abc import ABC, abstractmethod
+
+from ...utils import _ElementApiManager
+from .factory import _ClassApiFactory
+
+if t.TYPE_CHECKING:
+    from ...gui import Gui
+
+
+class ElementApi(ABC):
+    ELEMENT_NAME = ""
+
+    def __new__(cls, *args, **kwargs):
+        obj = super(ElementApi, cls).__new__(cls)
+        parent = _ElementApiManager().peek()
+        if parent is not None:
+            parent.add(obj)
+        return obj
+
+    def __init__(self, **kwargs):
+        self._properties = kwargs
+
+    def update(self, **kwargs):
+        self._properties.update(kwargs)
+
+    @abstractmethod
+    def _render(self, gui: "Gui") -> str:
+        pass
+
+
+class BlockElementApi(ElementApi):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._children: t.List[ElementApi] = []
+
+    def add(self, *elements: ElementApi):
+        for element in elements:
+            if element not in self._children:
+                self._children.append(element)
+        return self
+
+    def __enter__(self):
+        _ElementApiManager().push(self)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        _ElementApiManager().pop()
+
+    def _render(self, gui: "Gui") -> str:
+        el = _ClassApiFactory.create_element(gui, self.ELEMENT_NAME, self._properties)
+        return f"{el[0]}{self._render_children(gui)}</{el[1]}>"
+
+    def _render_children(self, gui: "Gui") -> str:
+        return "\n".join([child._render(gui) for child in self._children])
+
+
+class ControlElementApi(ElementApi):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def _render(self, gui: "Gui") -> str:
+        el = _ClassApiFactory.create_element(gui, self.ELEMENT_NAME, self._properties)
+        return f"<div>{el[0]}</{el[1]}></div>"
```

Comment:
do we need key and css class  for the div ?

Reply:
at least key should be unique
What is used the markdown postproc is & static method of _Builder that you may be able to re-use ?
https://vscode.dev/github/Avaiga/taipy-gui/blob/develop/src/taipy/gui/renderers/builder.py#L169